File: java/SmartHomeSystem.java
package smarthome;

import java.util.*;
import java.time.*;

public class SmartHomeSystem {

    public enum Role { ADMIN, USER, GUEST }

    private Map<String, Device> devices = new LinkedHashMap<>();
    private Map<String, Scenario> scenarios = new LinkedHashMap<>();
    private List<Event> events = new ArrayList<>();
    private List<Notification> notifications = new ArrayList<>();

    private Scheduler scheduler = new Scheduler();
    private EnergyManager energyManager = new EnergyManager();

    private List<String> eventLog = new ArrayList<>();
    private List<String> errorLog = new ArrayList<>();

    private double totalEnergyUsage = 0.0;

    public void addDevice(Device d) {
        if (d == null) return;

        if (devices.containsKey(d.getId())) {
            String msg = "Device ID already exists: " + d.getId();
            System.out.println(msg);
            logError(msg);
            return;
        }

        devices.put(d.getId(), d);
        System.out.println("Добавлено устройство: " + d.getName());
    }

    public void removeDevice(String id) {
        devices.remove(id);
        System.out.println("Удалено: " + id);
    }

    public Collection<Device> getDevices() {
        return devices.values();
    }

    public List<Device> getDevicesList() {
        return new ArrayList<>(devices.values());
    }

    public Device findDeviceById(String id) {
        return devices.get(id);
    }

    public Device findDeviceByName(String name) {
        return devices.values()
                .stream()
                .filter(d -> d.getName().equals(name))
                .findFirst()
                .orElse(null);
    }

    public void addScenario(Scenario s) {
        scenarios.put(s.getId(), s);
        System.out.println("Сценарий добавлен: " + s.getName());
    }

    public Collection<Scenario> getScenarios() {
        return scenarios.values();
    }

    public Scenario getScenarioById(String id) {
        return scenarios.get(id);
    }

    public void triggerEvent(Event e) {
        events.add(e);
        String info = LocalDateTime.now() + " " + e.getInfo();
        eventLog.add(info);
        e.trigger(this);
    }

    public void sendNotification(Notification n) {
        notifications.add(n);
        n.send();
    }

    public Scheduler getScheduler() {
        return scheduler;
    }

    public EnergyManager getEnergyManager() {
        return energyManager;
    }

    public void showDevices() {
        if (devices.isEmpty()) {
            System.out.println("Устройств нет");
            return;
        }

        for (Device d : devices.values()) {
            System.out.println(d.toString());
        }
    }

    public void showScenarios() {
        if (scenarios.isEmpty()) {
            System.out.println("Сценариев нет");
            return;
        }

        for (Scenario s : scenarios.values()) {
            System.out.println(s.toString());
        }
    }

    public void showEventLog() {
        if (eventLog.isEmpty()) {
            System.out.println("Событий нет");
        } else {
            for (String e : eventLog) System.out.println(e);
        }
    }

    public void logError(String msg) {
        String rec = LocalDateTime.now() + " ERROR: " + msg;
        errorLog.add(rec);
    }

    public void showErrorLog() {
        if (errorLog.isEmpty()) System.out.println("Ошибок нет");
        else errorLog.forEach(System.out::println);
    }

    public void safeTurnOnDevice(String id) {
        Device d = findDeviceById(id);

        if (d == null) {
            String msg = "Не найдено устройство " + id;
            System.out.println(msg);
            logError(msg);
            return;
        }

        try {
            d.turnOn();
            totalEnergyUsage += d.getPowerUsage();
        } catch (Exception e) {
            logError("safeTurnOnDevice: " + e.toString());
        }
    }

    public void safeTurnOffDevice(String id) {
        Device d = findDeviceById(id);

        if (d == null) {
            String msg = "Не найдено устройство " + id;
            System.out.println(msg);
            logError(msg);
            return;
        }

        try {
            d.turnOff();
        } catch (Exception e) {
            logError("safeTurnOffDevice: " + e.toString());
        }
    }

    public double getTotalEnergyUsage() {
        return totalEnergyUsage;
    }

    public void sendNotificationToDeviceId(String deviceId, String msg) {
        try {
            Device d = findDeviceById(deviceId);

            if (d == null)
                throw new IllegalArgumentException("Device not found: " + deviceId);

            Notification n = new Notification(msg);
            n.sendToDevice(d);

            notifications.add(n);

        } catch (Exception e) {
            logError("sendNotificationToDeviceId: " + e.toString());
            System.out.println("Ошибка уведомления: " + e.getMessage());
        }
    }

    public void showNotifications() {
        if (notifications.isEmpty())
            System.out.println("Уведомлений нет");
        else
            notifications.forEach(Notification::show);
    }

    public Optional<Scenario> safeFindScenario(String id) {
        return Optional.ofNullable(scenarios.get(id));
    }

    public void registerTrigger(String deviceId, java.util.function.Consumer<AbstractDevice> cb) {
        Device d = findDeviceById(deviceId);

        if (d == null) {
            logError("registerTrigger: device not found " + deviceId);
            return;
        }

        if (d instanceof AbstractDevice)
            ((AbstractDevice) d).addListener(cb);
    }
}

File: smarthome/shc/src/SmartHomeSystem.cpp
#include "SmartHomeSystem.hpp"

SmartHomeSystem::SmartHomeSystem() {}
SmartHomeSystem::~SmartHomeSystem() {}

void SmartHomeSystem::addDevice(std::shared_ptr<Device> d) {
    if (d) devices.push_back(d);
}

void SmartHomeSystem::removeDevice(std::shared_ptr<Device> d) {
    devices.erase(std::remove(devices.begin(), devices.end(), d), devices.end());
}

std::shared_ptr<Device> SmartHomeSystem::findDeviceByName(const std::string &name) {
    for (auto d: devices) if (d && d->getName() == name) return d;
    return nullptr;
}

void SmartHomeSystem::runDemo() {
    for (auto d: devices) if (d) d->showStatus();
    scheduler.runPending();
}

void SmartHomeSystem::addAction(std::shared_ptr<Action> a) {
    if (a) actions.push_back(a);
}

std::shared_ptr<Action> SmartHomeSystem::findAction(const std::string &name) const {
    for (auto a: actions) if (a && a->getName() == name) return a;
    return nullptr;
}

void SmartHomeSystem::addScenario(std::shared_ptr<Scenario> s) {
    if (s) scenarios.push_back(s);
}

std::shared_ptr<Scenario> SmartHomeSystem::findScenario(const std::string &name) const {
    for (auto s: scenarios) if (s && s->getName() == name) return s;
    return nullptr;
}

void SmartHomeSystem::addUser(std::shared_ptr<User> u) {
    if (u) users.push_back(u);
}

std::shared_ptr<User> SmartHomeSystem::findUser(int id) const {
    for (auto u: users) if (u && u->getUserId() == id) return u;
    return nullptr;
}

void SmartHomeSystem::showDevices() const {
    std::cout << "\n=== Устройства ===" << std::endl;
    for (auto d: devices) if (d) std::cout << "id=" << d->getId() << ", имя=" << d->getName() << ", локация=" << d->getLocation() << std::endl;
}

void SmartHomeSystem::showScenarios() const {
    std::cout << "\n=== Сценарии ===" << std::endl;
    for (auto s: scenarios) if (s) std::cout << s->getName() << std::endl;
}

void SmartHomeSystem::showActions() const {
    std::cout << "\n=== Действия ===" << std::endl;
    for (auto a: actions) if (a) std::cout << a->getName() << std::endl;
}

void SmartHomeSystem::sendNotificationToDevice(int deviceId, const std::string &msg) {
    auto n = std::make_shared<Notification>(msg);
    notifications.push_back(n);
    std::cout << "[Уведомление] устройству " << deviceId << ": " << msg << std::endl;
}

void SmartHomeSystem::sendNotificationToAll(const std::string &msg) {
    auto n = std::make_shared<Notification>(msg);
    notifications.push_back(n);
    std::cout << "[Уведомление] ВСЕМ: " << msg << std::endl;
}

void SmartHomeSystem::showNotifications() const {
    std::cout << "\n=== Журнал уведомлений ===" << std::endl;
    for (auto n: notifications) if (n) n->show();
}

EnergyManager& SmartHomeSystem::getEnergyManager() {
    return energyManager;
}

Scheduler& SmartHomeSystem::getScheduler() {
    return scheduler;
}

std::shared_ptr<Scenario> SmartHomeSystem::duplicateScenario(const std::string& existingName,
                                                             const std::string& newName,
                                                             bool deepCopy)
{
    auto s = findScenario(existingName);
    if (!s) return nullptr;

    auto copy = deepCopy ? s->cloneDeep() : s->cloneShallow();
    copy->adaptTo(newName);
    scenarios.push_back(copy);
    std::cout << "Сценарий '" << existingName << "' скопирован как '" << newName << "' (глубокое=" << deepCopy << ")" << std::endl;
    return copy;
}